File: Ruby/Parte2.rb
require 'csv'
require 'wordcloud'
require 'chartkick'

# Leer los datos del archivo CSV
data = CSV.read('web_scraping_citas.csv', headers: true)

# Ordenar por número de citas y guardar los 10 artículos principales
top_articles = data.sort_by { |row| row['Citado por'].to_i }.reverse[0..9]

# Guardar los 10 artículos principales en un archivo CSV
CSV.open('top_10_articles.csv', 'w') do |csv|
  csv << ["Número de Citas", "Año de Publicación", "Título del Artículo", "Revista"]
  top_articles.each do |article|
    csv << [article['Citado por'], article['Year'], article['Title'], article['Journal']]
  end
end

# Búsqueda de palabras clave en los títulos
print "Ingresa una palabra clave: "


keyword = gets.chomp

# Filtrar títulos que contienen la palabra clave
titles_with_keyword = data.select { |row| row['Title'].downcase.include?(keyword.downcase) }

# Guardar los resultados en un archivo CSV
CSV.open('titles_with_keyword.csv', 'w') do |csv|
  csv << ["Título", "URL"]
  titles_with_keyword.each do |article|
    csv << [article['Title'], article['URL']]
  end
end

# Contar todos los autores
author_counts = Hash.new(0)

data.each do |row|
  authors = row['Autor'].split('-')
  authors.each do |author|
    author_counts[author.strip] += 1
  end
end

# Ordenar por el número de apariciones y guardar en un archivo CSV
sorted_authors = author_counts.sort_by { |_author, count| -count }

CSV.open('author_counts.csv', 'w') do |csv|
  csv << ["Autor", "Número de Apariciones"]
  sorted_authors.each do |author, count|
    csv << [author, count]
  end
end

# Crear una nube de palabras
word_cloud_data = sorted_authors.to_h
word_cloud = WordCloud.new(word_cloud_data)
word_cloud.to_file('word_cloud.png')

# Contar la cantidad de artículos por año y crear un gráfico de barras
articles_by_year = data.group_by { |row| row['Year'] }.transform_values(&:count)

bar_chart = Chartkick::BarChart.new(articles_by_year, xtitle: 'Año', ytitle: 'Número de Artículos', library: { animation: { startup: true, duration: 1000 } })
bar_chart.save('articles_by_year.html')

File: R/Parte2.r
# Cargar las bibliotecas necesarias
library(dplyr)

# Leer los datos obtenidos del web scraping
data <- read.csv("web_scraping_citas.csv")

# 1. Identificar los 10 artículos con más citas
top_10_cited <- data %>%
  arrange(desc(Cited_By)) %>%
  head(10)

write.csv(top_10_cited, "top_10_citados.csv", row.names = FALSE)

# 2. Búsqueda de palabras clave en títulos
keyword <- readline("Ingresa una palabra clave: ")

articles_with_keyword <- data %>%
  filter(grepl(keyword, tolower(Title)))

write.csv(articles_with_keyword, "articles_with_keyword.csv", row.names = FALSE)

# 3. Contar todos los autores
author_counts <- data %>%
  separate_rows(Author, sep = ",") %>%
  trimws() %>%
  count(Author, sort = TRUE)

write.csv(author_counts, "author_counts.csv", row.names = FALSE)

# 4. Visualización gráfica
# 4.1. Gráfico de nube de palabras
library(tm)
library(wordcloud)

# Crear un corpus de los títulos
corpus <- Corpus(VectorSource(data$Title))
corpus <- tm_map(corpus, content_transformer(tolower))
corpus <- tm_map(corpus, removePunctuation)
corpus <- tm_map(corpus, removeNumbers)
corpus <- tm_map(corpus, removeWords, stopwords("spanish"))
corpus <- tm_map(corpus, stripWhitespace)

# Crear una matriz término-documento
dtm <- DocumentTermMatrix(corpus)
word_freqs <- row_sums(as.matrix(dtm))
wordcloud(names(word_freqs), word_freqs, min.freq = 3, colors = brewer.pal(8, "Dark2"))

# 4.2. Gráfico de barras de la cantidad de artículos por año de publicación
data$Year <- as.integer(data$Year)
year_counts <- data %>%
  filter(!is.na(Year)) %>%
  count(Year) %>%
  arrange(Year)

barplot(year_counts$n, names.arg = year_counts$Year, xlab = "Año de Publicación", ylab = "Número de Artículos")